File: packages/enforcement_py/alloist_enforce/revocation_verify.py
"""Verify signed revocation payloads from WebSocket."""

from __future__ import annotations

import base64
import json
import time
from typing import Any

import httpx
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

REVOCATION_KID = "revocation"
_MAX_AGE_SECONDS = 120
_CACHE_TTL = 300  # 5 minutes
_cached_key: tuple[bytes, float] | None = None


def _base64url_to_bytes(b64url: str) -> bytes:
    """Decode base64url to raw bytes."""
    padding = 4 - len(b64url) % 4
    if padding != 4:
        b64url += "=" * padding
    return base64.urlsafe_b64decode(b64url)


def _canonical_payload(data: dict[str, Any]) -> bytes:
    """Canonical JSON for verification (sort keys, compact)."""
    return json.dumps(data, sort_keys=True, separators=(",", ":")).encode("utf-8")


def fetch_revocation_public_key(api_url: str) -> bytes | None:
    """Fetch revocation public key from GET /keys. Caches with TTL."""
    global _cached_key
    now = time.time()
    if _cached_key and (now - _cached_key[1]) < _CACHE_TTL:
        return _cached_key[0]

    base = api_url.rstrip("/").replace("ws://", "http://").replace("wss://", "https://")
    url = f"{base}/keys"
    try:
        with httpx.Client(timeout=10.0) as client:
            resp = client.get(url)
        if not resp.is_success:
            return _cached_key[0] if _cached_key else None
        data = resp.json()
        for key in data.get("keys", []):
            if key.get("kid") == REVOCATION_KID:
                x = key.get("x")
                if not x:
                    return None
                raw = _base64url_to_bytes(x)
                _cached_key = (raw, now)
                return raw
        return _cached_key[0] if _cached_key else None
    except Exception:
        return _cached_key[0] if _cached_key else None


def verify_revocation_payload(payload: dict[str, Any], public_key_bytes: bytes | None) -> bool:
    """
    Verify signed revocation payload.
    Returns True if signature valid and payload not expired.
    """
    if not public_key_bytes:
        return False
    signature_b64 = payload.get("signature")
    kid = payload.get("kid")
    if not signature_b64 or kid != REVOCATION_KID:
        return False

    ts_str = payload.get("ts")
    if not ts_str:
        return False
    try:
        from datetime import datetime, timezone
        ts = datetime.fromisoformat(ts_str.replace("Z", "+00:00"))
        age = (datetime.now(timezone.utc) - ts).total_seconds()
        if age < 0 or age > _MAX_AGE_SECONDS:
            return False
    except (ValueError, TypeError):
        return False

    verify_payload = {
        "token_id": payload.get("token_id"),
        "event": payload.get("event"),
        "ts": ts_str,
        "nonce": payload.get("nonce"),
    }
    if None in verify_payload.values():
        return False

    try:
        sig = base64.b64decode(signature_b64.encode("ascii"))
        public_key = Ed25519PublicKey.from_public_bytes(public_key_bytes)
        payload_bytes = _canonical_payload(verify_payload)
        public_key.verify(sig, payload_bytes)
        return True
    except Exception:
        return False

File: packages/enforcement/src/revocationVerify.js
'use strict';

const crypto = require('crypto');

const REVOCATION_KID = 'revocation';
const MAX_AGE_SECONDS = 120;
const CACHE_TTL_MS = 5 * 60 * 1000; // 5 minutes

let cachedKey = null;
let cachedAt = 0;

function canonicalPayload(data) {
  return JSON.stringify({
    event: data.event,
    nonce: data.nonce,
    token_id: data.token_id,
    ts: data.ts,
  });
}

async function fetchRevocationPublicKey(apiUrl) {
  const now = Date.now();
  if (cachedKey && now - cachedAt < CACHE_TTL_MS) {
    return cachedKey;
  }

  const base = apiUrl.replace(/\/$/, '').replace(/^ws:\/\//, 'http://').replace(/^wss:\/\//, 'https://');
  const url = `${base}/keys`;

  try {
    const res = await fetch(url);
    if (!res.ok) return cachedKey;

    const data = await res.json();
    for (const key of data.keys || []) {
      if (key.kid === REVOCATION_KID && key.x) {
        const keyObj = crypto.createPublicKey({
          key: { kty: 'OKP', crv: 'Ed25519', x: key.x },
          format: 'jwk',
        });
        cachedKey = keyObj;
        cachedAt = now;
        return keyObj;
      }
    }
    return cachedKey;
  } catch {
    return cachedKey;
  }
}

function verifyRevocationPayload(payload, publicKey) {
  if (!publicKey) return false;

  const signatureB64 = payload.signature;
  const kid = payload.kid;
  if (!signatureB64 || kid !== REVOCATION_KID) return false;

  const tsStr = payload.ts;
  if (!tsStr) return false;

  try {
    const ts = new Date(tsStr.replace('Z', '+00:00'));
    const age = (Date.now() - ts.getTime()) / 1000;
    if (age < 0 || age > MAX_AGE_SECONDS) return false;
  } catch {
    return false;
  }

  const verifyPayload = canonicalPayload(payload);
  const signature = Buffer.from(signatureB64, 'base64');

  try {
    return crypto.verify(null, Buffer.from(verifyPayload, 'utf8'), publicKey, signature);
  } catch {
    return false;
  }
}

module.exports = {
  fetchRevocationPublicKey,
  verifyRevocationPayload,
};
